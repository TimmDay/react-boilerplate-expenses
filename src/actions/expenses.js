// ACTION
import database from '../firebase/firebase';
import uuid from "uuid"; // not needed when using firebase


// gives us access to dispatch so we can use it inside inner func
// move defaults to async wrapper func
// firebase generates auto uuid
// returning that function only works because we set up redux thunk in the store
export const startAddExpense = (expenseData = {}) => {

    return (dispatch) => {
        // default values for object items for return object. destructured from expense data
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData; //defaults when arg has missing info

        const expense = {description, note, amount, createdAt};

        // db is the firebase connection. return allows us to chain the returned promise
        return database.ref('expenses').push(expense)
            .then((ref) => {
                dispatch(addExpense({ // cannot forget this, or redux store is out of date
                    id: ref.key,
                    ...expense
                }))
            });
    };
};
// dispatch(addExpense({ // the non-firebase way with uuid
//     id: uuid(),
//     ...expense
// }));

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});



export const startRemoveExpense = ({ id='' } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
            .then(() => {
                dispatch(removeExpense({id})); //update store
            })
    }
};
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});


// EDIT_EXPENSE updates is an object with updated fields for reducer
//update is like set, but doesnt overwrite the parts we dont touch
export const startEditExpense = (id='', updates={}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates)
            .then(() => {
                dispatch(editExpense(id,updates));
            })
    }
};
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});




// ACTIONS FOR FETCH FROM DB

//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// for persisting data saved from a previous session
export const startSetExpenses = () => {
    return (dispatch) => {
        // console.log(database.ref('expenses').once('value')); //todo
        //fetch all expense data once, from firebase. return for the promise

        return database.ref('expenses').once('value')
            .then((dataSnapshot) => {

                const dbExpenses = [];
                dataSnapshot.forEach((childSnapshot) => {
                    // console.log(childSnapshot.key); //todo
                    // console.log(childSnapshot.val()); //todo
                    dbExpenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    })
                });

                dispatch(setExpenses(dbExpenses)); //dispatch set expenses
            });
    }
};