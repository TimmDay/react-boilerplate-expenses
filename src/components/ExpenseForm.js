import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates'; // requires moment


// this one is reusable.
// for use in AddExpense and EditExpense

// form fields, validation and logic

export default class ExpenseForm extends React.Component {


    // when form submitted, we will send it off to redux
    // to either edit existing expense or create new one
    //note optional, number can be zero if they want. just REQUIRE desc
    // define onCHange otherwise read only inputs

    // for amount input, enforce format
    // - only numbers, one decimal, two num after
    // - use string and convert to num last
    // conditional logic in oncChnage handler


    constructor(props) {
        super(props);
        // local component state
        this.state = {
            description: props.existingExpense ? props.existingExpense.description : '',
            note: props.existingExpense ? props.existingExpense.note : '',
            amount: props.existingExpense ? (props.existingExpense.amount / 100).toString() : '',
            createdAt: props.existingExpense ? moment(props.existingExpense.createdAt) : moment(), // current time
            calendarFocused: false,
            error: ''
        };
    }


    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description: description }));
    };
    onNoteChange = (e) => {
        // e.persist();
        this.setState(() => ({ note: e.target.value}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount: amount }));
        }
    };


    onDateChange = (createdAt) => {
        if (createdAt) { //prevents user clearing the value
            // because if date is cleared, this func gets called with nothing
            // (we don't want to update this part of state to nothing)
            this.setState(() => ({ createdAt: createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(()=> ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault(); //otherwise browser does full page refresh

        if (!this.state.description || !this.state.amount) {
            // set error state equal 'please provide description and amount'
            this.setState(() => ({error: 'please provide description and amount'}))
        } else {
            this.setState(() => ({error: ''}));
            // the dispatch to store func, added as prop in AddExpense when form instantiated
            // * 100 to cents
            // javascript works in ms
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };


    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form
                    onSubmit={this.onSubmit}
                >
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />

                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />

                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false} //every day available, past and future
                    />

                    <textarea
                        placeholder="add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>

                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}
