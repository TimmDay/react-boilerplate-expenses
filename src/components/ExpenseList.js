import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from './../selectors/expenses';

// use map to iterate over list and render each one
// export unconnected verion (for testing only)
export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-larger-screens">Expense</div>
            <div className="show-for-larger-screens">Amount</div>
        </div>

        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div>
                        <span className="list-item list-item--empty-msg">no expenses</span>
                    </div>
                ) : (
                    props.expenses.map((ex) => (
                        <ExpenseListItem
                            key={ex.id}
                            {...ex}
                        />
                    ))
                )
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);

//state.expenses is the complete array. to filter it we need our selector

// HOC - to access redux store
// connect gives us a function back... so second parenthesis to use pass in component
// in the connect call we also specify what part of the store we want to access, with a func
// return an object


// long version of code, for clarity of understanding / reference
// const ConnectedExpenseList = connect((state) => {
//
//     return {
//         expenses : state.expenses
//     };
//
// })(ExpenseList);
//
//
// export default ConnectedExpenseList;

//alternative: export default connect ((....
