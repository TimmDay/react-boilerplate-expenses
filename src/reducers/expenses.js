// Expenses Reducer
const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            //code that adds expense to the array
            // do not change the original array, so no push
            // return state.concat(action.expense);
            return [
                ...state,
                action.expense
            ];

        case 'REMOVE_EXPENSE' :
            // const newArr = [];
            // state.forEach((item) => {
            //     if (item.id !== action.id){
            //         newArr.push(item);
            //     }
            // });
            // return newArr;

            return state.filter(({ id }) => id !== action.id );

        case 'EDIT_EXPENSE':
            // go through array for match
            // when found, correctly change it

            return state.map((ex) => {
                if (ex.id === action.id) {
                    //return edited object
                    return {
                        ...ex,
                        ...action.updates
                        // action updates is whatever was called in the dispatch
                        // will override
                    }

                } else {
                    return ex;
                }
            });

        case 'SET_EXPENSES':
            // remove existing expenses from store
            // add provided expenses to store
            return action.expenses;

        default:
            return state;
    }
};

export default expensesReducer;