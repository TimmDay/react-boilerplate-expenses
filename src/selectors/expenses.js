import moment from 'moment';

// GET visibile expenses
//combine info from the two stores to display only filtered data

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate }) => {

    return expenses.filter((expense) => {
        // if typeof not number, say true (so that it has no effect)
        // true items are not filtered
        // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;

        // console.log(moment());
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true; //never filter if no endDate

        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
            //most recent first. so the -1 goes to the highest number

        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
            // larget amount first
        }
    })
};

export default getVisibleExpenses;