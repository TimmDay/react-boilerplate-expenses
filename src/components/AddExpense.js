import React from 'react';
import {Link} from 'react-router-dom';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux'; //connect so we can dispatch to store

import {startAddExpense} from "../actions/expenses";

// export so that we can test the unconnected version of the component
export class AddExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">add expense</h1>
                    </div>
                </div>


                <div className="content-container">
                    <ExpenseForm
                        onSubmit={this.onSubmit}
                    />
                </div>

                <Link to="/dashboard">back</Link>
            </div>
        )
    }
}


const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpense);

// first parenthesis for mapping state as props, 2nd the component
// with connect wired up, we can access props.dispatch

// components rendered inside react router get access to special props -> history.push
// push takes single string path arg
// switches over like link click (no full page refresh)