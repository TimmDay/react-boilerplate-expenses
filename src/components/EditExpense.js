import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from "../actions/expenses";


export class EditExpense extends React.Component {

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/dashboard');
    };

    onRemove = () => {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                {/*<p>editing expense {this.props.match.params.id} </p>*/}

                {/* populate fields with existing values */}
                <ExpenseForm
                    existingExpense={this.props.expense}
                    onSubmit={this.onSubmit}
                />

                <button
                    onClick = {this.onRemove}
                >Remove</button>
            </div>
        )
    }
}

// give the component the current expense object as props
const mapStateToProps = (state, props) =>
{
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
};
// MORE CONCISE PATTERN OF ABOVE
//     ({
//     expense: state.expenses.find((expense) => {
//         return expense.id === props.match.params.id;
//     })
// });

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);