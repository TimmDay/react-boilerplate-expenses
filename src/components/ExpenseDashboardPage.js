import React from 'react';
import { Link } from 'react-router-dom';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
    <div>
        <p> the dashboard component </p>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;