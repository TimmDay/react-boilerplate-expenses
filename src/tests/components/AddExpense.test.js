import React from 'react';
import { shallow } from 'enzyme';
import { AddExpense } from "../../components/AddExpense";
import expenses from '../fixtures/expenses';


// build the test stuff before each test case. reuse spies and wrappers for eac test
let startAddExpense, history, wrapper;
beforeEach(() => {
    startAddExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpense startAddExpense={startAddExpense} history={history}/>);
});


// the comp needs two props
test('should render add expense correctly', () => {
    expect(wrapper).toMatchSnapshot();
});


test('should handle onSubmit', () => {
    // const onSubmit = jest.fn();
    // const history = { push: jest.fn() };
    // const wrapper = shallow(<AddExpense startAddExpense={onSubmit} history={history}/>);
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]); //call it with an expense data to submit
    expect(history.push).toHaveBeenLastCalledWith('/dashboard');
    expect(startAddExpense).toHaveBeenLastCalledWith(expenses[1]);

});