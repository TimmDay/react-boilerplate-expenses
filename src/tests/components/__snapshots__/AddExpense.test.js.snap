// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expense correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        add expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
  <Link
    replace={false}
    to="/dashboard"
  >
    back
  </Link>
</div>
`;
